### PR TITLE
Be able to edit a node source through the CLI

### PR DESCRIPTION
**Pull request by jrochas**
Introduce a new CLI command named 'editns'. It allows to change the parameters of an undeployed node source to be taken into account in the next deployment.

The 'editns' command fails if the node source is currently deployed.

The parameters to give to this command are the same as for node source definition/creation, but all parameters are mandatory.

### CODE REVIEW

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/CommandSet.java [hunk 1]
```diff
@@ -967,7 +979,7 @@
      */
     public static final CommandSet.Entry[] RM_ONLY = new CommandSet.Entry[] { LOGIN, NODE_ADD, NODE_LIST, NODE_INFO,
                                                                               NODE_LOCK, NODE_UNLOCK, NODE_REMOVE,
-                                                                              NS_DEFINE, NS_CREATE, NS_DEPLOY,
+                                                                              NS_DEFINE, NS_EDIT, NS_CREATE, NS_DEPLOY,
```
Comment (sonarqube-activeeon) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/CommandSet.java]: MINOR Remove this use of "NS_CREATE"; it is deprecated. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/rm/EditNodeSourceCommand.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive_grid_cloud_portal.cli.cmd.rm;
```
Comment (sonarqube-activeeon) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/rm/EditNodeSourceCommand.java]: MINOR Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00120)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java [hunk 3]
```diff
@@ -130,95 +150,137 @@ private void defineNodeSourceAndListNodeSources(String nodeSourceName, boolean n
...
+        this.checkOccurrencesOfNodeSourceInNodeSourceList(nodeSourceName);
+        this.checkNoOccurrenceOfNodeSourceInNodesList(nodeSourceName);
     }
 
-    private void checkOccurrencesOfNodeSourceInNodeSourceList(String testLogString, String nodeSourceName) {
-        System.out.println(testLogString + " List node sources");
+    private void editNodeSourceAndCheckOutput(String nodeSourceName) {
+
+        String nodeSourceInfrastructureClass = "org.ow2.proactive.resourcemanager.nodesource.infrastructure.LocalInfrastructure";
+        String nodeSourcePolicyClass = "org.ow2.proactive.resourcemanager.nodesource.policy.StaticPolicy";
```
Comment (Mykhailenko) [inline rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java]: why not this `String nodeSourcePoliceClass = StaticPolicy.getClass().getName()` ?

File: rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java [hunk 4]
```diff
@@ -130,95 +150,137 @@ private void defineNodeSourceAndListNodeSources(String nodeSourceName, boolean n
...
+
+        System.out.println(LOG_HEADER + " Test editns command");
+
+        this.clearAndTypeLine("editns( '" + nodeSourceName + "', ['" + nodeSourceInfrastructureClass + "', '" +
+                              rmCredentialPath + "', " + MODIFIED_NB_NODES + ", 60000, ''], ['" +
+                              nodeSourcePolicyClass + "', 'ALL', 'ALL'], 'false')");
```
Comment (Mykhailenko) [inline rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java]: This is not so easy to read.
Maybe you can use String.format?

File: rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java (hunk 3 above)
Comment (jrochas) [reply rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java]: That makes sense. Thank you

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/rm/EditNodeSourceCommand.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
+/*
+ * ProActive Parallel Suite(TM):
+ * The Open Source library for parallel and distributed
+ * Workflows & Scheduling, Orchestration, Cloud Automation
+ * and Big Data Analysis on Enterprise Grids & Clouds.
+ *
+ * Copyright (c) 2007 - 2017 ActiveEon
```
Comment (Mykhailenko) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/rm/EditNodeSourceCommand.java]: Could you change to 2018?

File: rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java (hunk 3 above)
Comment (fviale) [reply rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java]: I prefer getCanonicalName()

File: rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java (hunk 4 above)
Comment (jrochas) [reply rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java]: I created methods to generate the infrastructure string and the policy string.

File: rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java (hunk 3 above)
Comment (jrochas) [reply rest/rest-cli/src/test/java/functionaltests/NodeSourceCommandsFunctTest.java]: Done.

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/rm/EditNodeSourceCommand.java (hunk 5 above)
Comment (jrochas) [reply rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/rm/EditNodeSourceCommand.java]: I am not sure for now.

**Pull request merged**